src/libsync/abstractnetworkjob.cpp

```diff
@@ -175,6 +175,13 @@ void AbstractNetworkJob::slotFinished()
     if (_followRedirects && !redirectUrl.isEmpty()) {
         _redirectCount++;
 
+        // For POSTs where the query args are sent in the body, we need to
+        // manually forward these to the target url - the server can only
+        // return the query-arg-less redirected url in that case.
+        if (reply()->operation() == QNetworkAccessManager::PostOperation && !_requestBody) {
```

Comment:
yes.
That's kind of a hack, i fgound myself surpsie that Qt does change the query url to the post data. 
So that's two hacks combinated.

As discussed on IRC, my original intention was that the query URL stay in the URL, as the README of https://github.com/owncloud/oauth2 mention, i did not expect it to be moved to the POST data. So I think the best would be to put them directly in the data and not in the query URL. 

Maybe there should be an extra check here to check that redirectUrl does not already have an encoded query?

Reply:
Good point about not overwriting any query args from the forward. And I might also change the existing POST call to just have a body instead.